server/tests/redis_test.rs

```diff
@@ -1,54 +1,242 @@
-// use std::fs;
+use std::str::FromStr;
 
-// use redis::{Client, Commands};
-// use testcontainers::{clients::Cli, images::redis::Redis, Container};
-// use unleash_edge::{
-//     data_sources::redis_provider::{RedisProvider, FEATURE_PREFIX, TOKENS_KEY},
-//     types::{EdgeProvider, EdgeToken},
-// };
+use redis::{Client, Commands};
+use testcontainers::{clients::Cli, images::redis::Redis, Container};
+use tokio::sync::mpsc;
 
-// const TOKEN: &str = "*:development.03fa5f506428fe80ed5640c351c7232e38940814d2923b08f5c05fa7";
+use unleash_edge::{
+    data_sources::redis_provider::{RedisProvider, FEATURE_PREFIX},
+    types::{EdgeSink, EdgeSource, EdgeToken, TokenValidationStatus},
+};
+use unleash_types::client_features::{ClientFeature, ClientFeatures};
 
-// fn setup_redis(docker: &Cli) -> (Client, String, Container<Redis>) {
-//     let node: Container<Redis> = docker.run(Redis::default());
-//     let host_port = node.get_host_port_ipv4(6379);
-//     let url = format!("redis://127.0.0.1:{host_port}");
+const TOKEN: &str = "*:development.03fa5f506428fe80ed5640c351c7232e38940814d2923b08f5c05fa7";
 
-//     (redis::Client::open(url.clone()).unwrap(), url, node)
-// }
+fn setup_redis(docker: &Cli) -> (Client, String, Container<Redis>) {
+    let node: Container<Redis> = docker.run(Redis::default());
+    let host_port = node.get_host_port_ipv4(6379);
+    let url = format!("redis://127.0.0.1:{host_port}");
 
-// #[tokio::test]
-// async fn redis_provider_returns_expected_data() {
-//     let docker = Cli::default();
-//     let (mut client, url, _node) = setup_redis(&docker);
+    (redis::Client::open(url.clone()).unwrap(), url, node)
+}
 
-//     let content = fs::read_to_string("../examples/features.json").expect("Could not read file");
+fn build_features_key(token: &EdgeToken) -> String {
+    token
+        .environment
+        .as_ref()
+        .map(|environment| format!("{FEATURE_PREFIX}{environment}"))
+        .expect("Tying to resolve features for a token that hasn't been validated")
+}
```

Comment:
Should we use the function from redis_provider here? Maybe it's a thing about function scopes and visibility that makes it difficult to do that

Reply:
I intentionally didn't do that because I don't want to expose that function. I figure it's okay because if you'll break these tests hard if the backing implementation changes. Don't mind a bit of duplication in tests to preserve sanity in the public API